_posts/2019-06-07-python-datascience-with-openfaas.md

```diff
@@ -0,0 +1,267 @@
+```
+
+### Loading the model into function memory
+Using the `HTTP` mode in the [of-watchdog][of-watchdog-homepage] enables us to load this model into memory once and reuse it for multiple requests. We only need to load the model at the start the `handler.py` implementation
+
+```python
+import json
+import os
+from pathlib import PurePath
+from typing import Any, List
+
+import torch
+from torch.autograd import Variable
+
+from .core import const, model, utils
+
+FUNCTION_ROOT = os.environ.get("function_root", "/root/function/")
+
+# init model
+RNN = model.RNN(const.N_LETTERS, const.N_HIDDEN, const.N_CATEGORIES)
+# fill in weights
+RNN.load_state_dict(
+    torch.load(str(PurePath(FUNCTION_ROOT, "data/char-rnn-classification.pt")))
+)
+
+
+def predict(line: str, n_predictions: int = 3) -> List[Any]:
+    """omitted for bevity"""
+
+def handle(req: bytes) -> str:
+    """handle a request to the function
+    Args:
+        req (bytes): request body
+    """
+
+    if not req:
+        return json.dumps({"error": "No input provided", "code": 400})
+
+    name = str(req)
+    output = predict(name)
+
+    return json.dumps(output)
+```
+
+When a function instance is started, e.g. during scaling or deployment, this file is parsed and the model is loaded into memory. This will happen exactly once because the http mode loads this handler as a very small background web sever.  The original forking mode in the watchdog would instead load this file for every invocation. When loading models, this creates an unacceptable latency.
+
+This template uses Flask to power the background web server. The `handle` can return any value that [Flask would accept][flask-return-values].  This provides a lot of flexibility to control the response. For example, you could instead set the error status code and body like this
+
+```python
+return json.dumps({"error": "No input provided"}), 400
+```
+
+### Deployment
+This template is designed to bundle the pre-trained model into the final docker image. This means that the deployment steps look like
+
+```bash
+cd classify && python train.py && cd -
+faas-cli build classify --tag=sha
+```
+
+This results in a completely self-contained docker image that does not need access to a database or S3 during runtime. This provides several benefits:
+
+1. This model becomes very easy to share because you don't need to share access to your model storage.
+2. It also means that each new deployment is completely versioned, you know the version of the model from the docker tag. This makes rolling backward or forward as simple as changing the docker tag on the deployment.
+3. One final benefit is that startup time can be slightly faster because once a node has the docker image in its cache, it has everything it needs to start new instances of the function.  If you load the model from an external source, then every function instance that starts _must_ curl/copy from that external location, there is no way to cache it for new instances.
+
```

Comment:
Any downsides?  Larger images?

Reply:
In most cases, models aren't actually that large.  For example, installing Pandas adds more to the final image size than the trained model, it is only 107K.   I will add a note that this does result in a larger image.  

The other downside is that deploying a new version of the model does require a new docker image build.  When models are deployed to something like a bucket, the function can be updated by only copying the model.  _But_ i would argue that the tooling for CI/CD of a docker image is simpler and much more well established compared to trying to dynamically update the model from a bucket.  Dynamic update from a bucket requires custom code in each function.

Reply:
I have added a comment

Reply:
curl -> `curl`